library/src/main/java/com/joom/grip/CombinedGripFactory.kt

```diff
@@ -0,0 +1,123 @@
+import com.joom.grip.mirrors.ClassMirror
+import com.joom.grip.mirrors.Type
+import java.nio.file.Path
+import java.util.concurrent.ConcurrentHashMap
+
+interface CombinedGripFactory {
+  fun create(grip: Grip, vararg grips: Grip): Grip
+  fun create(grips: Iterable<Grip>): Grip
+
+  companion object {
+    @JvmStatic
+    val INSTANCE: CombinedGripFactory = CombinedGripFactoryImpl()
+  }
+}
+
+private class CombinedGripFactoryImpl : CombinedGripFactory {
+  override fun create(grips: Iterable<Grip>): Grip {
+    val classRegistry = CombinedClassRegistryImpl(grips)
+    val fileRegistry = CombinedFileRegistryImpl(grips.map { it.fileRegistry })
+
+    return GripImpl(fileRegistry, classRegistry, grips)
+  }
+
+  override fun create(grip: Grip, vararg grips: Grip): Grip {
+    return create(listOf(grip) + grips.asList())
+  }
+}
+
+private class CombinedClassRegistryImpl(private val grips: Iterable<Grip>) : ClassRegistry {
+  private val annotationTypeToGrip = ConcurrentHashMap<Type.Object, Grip>()
+  private val typeToGrip = ConcurrentHashMap<Type.Object, Grip>()
+  override fun getAnnotationMirror(type: Type.Object): AnnotationMirror {
+    val grip = annotationTypeToGrip.getOrPut(type) {
+      grips.first { it.fileRegistry.contains(type) }
+    }
+
+    return grip.classRegistry.getAnnotationMirror(type)
+  }
+
+  override fun getClassMirror(type: Type.Object): ClassMirror {
+    val grip = typeToGrip.getOrPut(type) {
+      grips.first { it.fileRegistry.contains(type) }
+    }
+
+    return grip.classRegistry.getClassMirror(type)
+  }
+}
+
+private class CombinedFileRegistryImpl(private val registries: Iterable<FileRegistry>) : FileRegistry {
+  private val typeToRegistry = ConcurrentHashMap<Type.Object, FileRegistry>()
+  private val pathToRegistry = ConcurrentHashMap<Path, FileRegistry>()
+
+  private val classpath by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
+    registries.flatMap { it.classpath() }
+  }
+
+  override fun classpath(): Collection<Path> {
+    return classpath
+  }
+
+  override fun readClass(type: Type.Object): ByteArray {
+    return registryByType(type)?.readClass(type) ?: run {
+      throw IllegalArgumentException("Unable to find registry for ${type.internalName}")
+    }
+  }
+
+  override fun contains(path: Path): Boolean {
+    return registryByPath(path) != null
+  }
+
+  override fun contains(type: Type.Object): Boolean {
+    return registryByType(type) != null
+  }
+
+  override fun findPathForType(type: Type.Object): Path? {
+    return registryByType(type)?.findPathForType(type)
+  }
+
+  override fun findTypesForPath(path: Path): Collection<Type.Object> {
+    return registryByPath(path)?.findTypesForPath(path) ?: run {
+      throw IllegalArgumentException("File $path is not added to the registry")
+    }
+  }
+
+  private fun registryByType(type: Type.Object): FileRegistry? {
+    typeToRegistry[type]?.let {
+      return it
+    }
+
+    return registries.firstOrNull { it.contains(type) }?.also {
```

Comment:
Is it necessary to check that several FileRegistry-instances matches given type or path? Does it matter?

Reply:
You mean that we should throw an error when several `FileRegistry` instances matches given type or path?

Reply:
Yep. But I'm not sure that it's really invalid state

Reply:
Well, according to original `FileRegistryImpl` it should not be possible, hence `Type` or `Path`serve as keys in `Map`. I will add a check for this